shim/src/functions.js

```diff
@@ -0,0 +1,59 @@
+// Adapted from SES/Caja
+// Copyright (C) 2011 Google Inc.
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/startSES.js
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/repairES5.js
+
+import { defineProperty, defineProperties, getPrototypeOf, setPrototypeOf } from './commons';
+
+/**
+ * The process to repair constructors:
+ * 1. Obtain the prototype from an instance
+ * 2. Create a substitute noop constructor
+ * 3. Replace its prototype property with the original prototype
+ * 4. Replace its prototype property's constructor with itself
+ * 5. Replace its [[Prototype]] slot with the noop constructor of Function
+ */
+function repairFunction(sandbox, functionName, functionDecl) {
+  const { unsafeEval, unsafeFunction } = sandbox;
+
+  const FunctionInstance = unsafeEval(`(${functionDecl}(){})`);
+  const FunctionPrototype = getPrototypeOf(FunctionInstance);
+
+  // Block evaluation of source when calling constructor on the prototype of functions.
+  const TamedFunction = unsafeFunction('throw new Error();');
+
+  defineProperties(TamedFunction, {
+    name: {
+      value: functionName
+    },
+    prototype: {
+      value: FunctionPrototype
+    }
+  });
+  defineProperty(FunctionPrototype, 'constructor', { value: TamedFunction });
+
+  // Prevent loop in case of Function.
+  if (functionName !== 'Function') {
+    setPrototypeOf(TamedFunction, unsafeFunction.prototype.constructor);
+  }
+}
+
+/**
+ * This block replaces the original Function constructor, and the original
+ * %GeneratorFunction% %AsyncFunction% and %AsyncGeneratorFunction%, with
```

Comment:
Note that this is another example of where a shim as normal user code cannot be safe against future specs, as those future specs may add more function constructors. This adds to the motivation for Realms to become standard